tracing/src/main/java/com/palantir/remoting1/tracing/AsyncSlf4jSpanObserver.java

```diff
@@ -32,34 +38,63 @@
+    @JsonSerialize(as = ImmutableZipkinCompatSpan.class)
     @Value.Immutable
     @Value.Style(visibility = Value.Style.ImplementationVisibility.PACKAGE)
-    abstract static class ZipkinCompatibleSerializableSpan {
-        private static final ObjectMapper mapper = new ObjectMapper().registerModule(new GuavaModule());
+    abstract static class ZipkinCompatSpan {
 
         abstract String getTraceId();
         abstract String getId();
         abstract String getName();
         abstract Optional<String> getParentId();
         abstract long getTimestamp();
         abstract long getDuration();
+        abstract List<ZipkinCompatAnnotation> annotations();
 
-        static ZipkinCompatibleSerializableSpan fromSpan(Span span) {
-            return ImmutableZipkinCompatibleSerializableSpan.builder()
+        static ZipkinCompatSpan fromSpan(Span span, ZipkinCompatEndpoint endpoint) {
+            return ImmutableZipkinCompatSpan.builder()
                     .traceId(span.getTraceId())
                     .id(span.getSpanId())
                     .name(span.getOperation())
                     .parentId(span.getParentSpanId())
                     .timestamp(span.getStartTimeMicroSeconds())
                     .duration(nanoToMicro(span.getDurationNanoSeconds()))  // Zipkin-durations are micro-seconds, round
+                    .addAllAnnotations(spanTypeToZipkinAnnotations(span, endpoint))
                     .build();
         }
 
+        private static Iterable<? extends ZipkinCompatAnnotation> spanTypeToZipkinAnnotations(
+                Span span, ZipkinCompatEndpoint endpoint) {
+            if (!span.type().isPresent()) {
+                return ImmutableList.of();
+            }
+
+            List<ZipkinCompatAnnotation> annotations = Lists.newArrayListWithCapacity(2);
+            switch (span.type().get()) {
+                case CLIENT_OUTGOING:
+                    annotations.add(ZipkinCompatAnnotation.of(span.getStartTimeMicroSeconds(), "cs", endpoint));
+                    annotations.add(ZipkinCompatAnnotation.of(
+                            span.getStartTimeMicroSeconds() + nanoToMicro(span.getDurationNanoSeconds()),
+                            "cr",
+                            endpoint));
+                    break;
+                case SERVER_INCOMING:
+                    annotations.add(ZipkinCompatAnnotation.of(span.getStartTimeMicroSeconds(), "sr", endpoint));
+                    annotations.add(ZipkinCompatAnnotation.of(
+                            span.getStartTimeMicroSeconds() + nanoToMicro(span.getDurationNanoSeconds()),
+                            "ss",
+                            endpoint));
+                    break;
+                default: // empty
+            }
+            return annotations;
+        }
+
         static long nanoToMicro(long nano) {
             return Math.max(1, (long) Math.ceil(nano / (double) 1000));
```

Comment:
Doing this via `double` math is still 3x slower than `return (nano + 1000) / 1000L;`